framework/src/timesteppers/CompositionDT.C

```diff
@@ -60,40 +64,100 @@ CompositionDT::CompositionDT(const InputParameters & parameters)
           "lower_bound", "Failed to find a timestepper with the name '", time_stepper_name, "'");
 }
 
+template <typename Lambda>
+void
+CompositionDT::actOnTimeSteppers(Lambda && act)
+{
+  for (auto & ts : getTimeSteppers())
+    act(*ts);
+}
+
+void
+CompositionDT::init()
+{
+  actOnTimeSteppers([](auto & ts) { ts.init(); });
+}
+
+void
+CompositionDT::preExecute()
+{
+  actOnTimeSteppers([](auto & ts) { ts.preExecute(); });
+}
+
+void
+CompositionDT::preSolve()
+{
+  actOnTimeSteppers([](auto & ts) { ts.preSolve(); });
+}
+
+void
+CompositionDT::postSolve()
+{
+  actOnTimeSteppers([](auto & ts) { ts.postSolve(); });
+}
+
+void
+CompositionDT::postExecute()
+{
+  actOnTimeSteppers([](auto & ts) { ts.postExecute(); });
+}
+
+void
+CompositionDT::preStep()
+{
+  actOnTimeSteppers([](auto & ts) { ts.preStep(); });
+}
+
+void
+CompositionDT::postStep()
+{
+  actOnTimeSteppers([](auto & ts) { ts.postStep(); });
+}
+
+bool
+CompositionDT::constrainStep(Real & dt)
+{
+  bool at_sync_point = TimeStepper::constrainStep(dt);
+  const auto time_steppers = getTimeSteppers();
+  for (auto & ts : time_steppers)
+    if (!ts->constrainStep(dt))
+      return false;
```

Comment:
so if all time steppers do not constrain, then we dont.

shouldnt it be "if any TS constrains, we constrain" ?